### PR TITLE
test: fix flaky dispatch loop

### LINKED ISSUE #50409: panic in the TestManagerDispatchLoop
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 3. What did you see instead (Required)

```
 [2024/01/15 03:03:17.975 +00:00] [ERROR] [spool.go:139] ["recover panic"] [recover="runtime error: index out of range [2227] with length 6"] [stack="github.com/pingcap/tidb/pkg/resourcemanager/pool/spool.(*Pool).run.func1.1\n\tpkg/resourcemanager/pool/spool/spool.go:139\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:914\nruntime.goPanicIndex\n\tGOROOT/src/runtime/panic.go:114\ngithub.com/pingcap/tidb/pkg/disttask/framework/scheduler_test.TestManagerDispatchLoop.func3.1\n\tpkg/disttask/framework/scheduler/scheduler_test.go:550\nreflect.Value.call\n\tGOROOT/src/reflect/value.go:596\nreflect.Value.Call\n\tGOROOT/src/reflect/value.go:380\ngo.uber.org/mock/gomock.(*Call).Do.func1\n\texternal/org_uber_go_mock/gomock/call.go:179\ngo.uber.org/mock/gomock.(*Controller).Call\n\texternal/org_uber_go_mock/gomock/controller.go:229\ngithub.com/pingcap/tidb/pkg/disttask/framework/mock.(*MockScheduler).ScheduleTask\n\tpkg/disttask/framework/mock/scheduler_mock.go:171\ngithub.com/pingcap/tidb/pkg/disttask/framework/scheduler.(*Manager).startScheduler.func1\n\tpkg/disttask/framework/scheduler/scheduler_manager.go:321\ngithub.com/pingcap/tidb/pkg/resourcemanager/pool/spool.(*Pool).run.func1\n\tpkg/resourcemanager/pool/spool/spool.go:144"] [stack="github.com/pingcap/tidb/pkg/resourcemanager/pool/spool.(*Pool).run.func1.1\n\tpkg/resourcemanager/pool/spool/spool.go:139\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:914\nruntime.goPanicIndex\n\tGOROOT/src/runtime/panic.go:114\ngithub.com/pingcap/tidb/pkg/disttask/framework/scheduler_test.TestManagerDispatchLoop.func3.1\n\tpkg/disttask/framework/scheduler/scheduler_test.go:550\nreflect.Value.call\n\tGOROOT/src/reflect/value.go:596\nreflect.Value.Call\n\tGOROOT/src/reflect/value.go:380\ngo.uber.org/mock/gomock.(*Call).Do.func1\n\texternal/org_uber_go_mock/gomock/call.go:179\ngo.uber.org/mock/gomock.(*Controller).Call\n\texternal/org_uber_go_mock/gomock/controller.go:229\ngithub.com/pingcap/tidb/pkg/disttask/framework/mock.(*MockScheduler).ScheduleTask\n\tpkg/disttask/framework/mock/scheduler_mock.go:171\ngithub.com/pingcap/tidb/pkg/disttask/framework/scheduler.(*Manager).startScheduler.func1\n\tpkg/disttask/framework/scheduler/scheduler_manager.go:321\ngithub.com/pingcap/tidb/pkg/resourcemanager/pool/spool.(*Pool).run.func1\n\tpkg/resourcemanager/pool/spool/spool.go:144"]
    scheduler_test.go:602: 
        	Error Trace:	pkg/disttask/framework/scheduler/scheduler_test.go:602
        	Error:      	Condition never satisfied
        	Test:       	TestManagerDispatchLoop 
```

https://tiprow.hawkingrei.com/view/gs/pingcapprow/pr-logs/pull/pingcap_tidb/50405/tiprow_fast_test/1746727973362339840#1:build-log.txt%3A4492

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: close #50409

Problem Summary:

### What changed and how does it work?
Don't start scheduler if it start time(counter) > len(concurrencies)

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by hawkingrei — APPROVED** [review]

File: pkg/disttask/framework/scheduler/scheduler_test.go [hunk 1]
```diff
@@ -535,6 +535,9 @@ func TestManagerDispatchLoop(t *testing.T) {
 	scheduler.RegisterSchedulerFactory(proto.TaskTypeExample,
 		func(ctx context.Context, task *proto.Task, param scheduler.Param) scheduler.Scheduler {
 			idx := counter.Load()
+			if int(idx) > len(concurrencies) {
```
Comment (D3Hunter) [inline pkg/disttask/framework/scheduler/scheduler_test.go:538]: might panic again when return nil

Comment (D3Hunter) [reply pkg/disttask/framework/scheduler/scheduler_test.go:538]: we should add a check in `mockScheduler.EXPECT().ScheduleTask().Do(func() {`, when task already success, just return

File: pkg/disttask/framework/scheduler/scheduler_test.go [hunk 2]
```diff
@@ -527,35 +526,34 @@ func TestManagerDispatchLoop(t *testing.T) {
...
 				require.NoError(t, taskMgr.WithNewSession(func(se sessionctx.Context) error {
 					_, err := sqlexec.ExecSQL(ctx, se, "update mysql.tidb_global_task set state=%?, step=%? where id=%?",
 						proto.TaskStateSucceed, proto.StepDone, task.ID)
 					return err
 				}))
+
 			})
 			mockScheduler.EXPECT().Close()
-			counter.Add(1)
+			waitChannels[task.Key] = make(chan struct{})
```
Comment (D3Hunter) [inline pkg/disttask/framework/scheduler/scheduler_test.go:556]: init this map before create task, else there might be data race

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**